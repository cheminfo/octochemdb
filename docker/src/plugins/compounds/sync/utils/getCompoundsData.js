import delay from 'delay';
import OCL from 'openchemlib';

import debugLibrary from '../../../../utils/Debug.js';

const debug = debugLibrary('getCompoundsData');
// node debug
/**
 * @description Calculate compounds properties (e.g. charge, OCL ID, molecular formula, etc.)
 * @param {*} molecule molecule from pubchem file
 * @returns  compounds properties
 */
export async function getCompoundsData(molecule, options = {}) {
  //console.log(OCL.Molecule.fromIDCode(molecule.idCode));

  if (!('indexRequired' in options)) {
    options.indexRequired = true;
  }
  try {
    let oclMolecule;
    if (molecule.molfile) {
      oclMolecule = OCL.Molecule.fromMolfile(molecule.molfile);
    } else if (molecule.smiles) {
      oclMolecule = OCL.Molecule.fromSmiles(molecule.smiles);
    } else {
      oclMolecule = OCL.Molecule.fromIDCode(molecule.idCode);
    }
    const oclID = oclMolecule.getIDCodeAndCoordinates();

    let urlIDCode = encodeURIComponent(oclID.idCode);
    let success = false;
    let count = 0;
    let dataCompound;
    while (success === false && count < 3) {
      try {
        // workerpool does not access the .env file for some reason, this is a workaround
        const controller = new AbortController();
        setTimeout(() => controller.abort(), 1000 * 1800);

        dataCompound = await fetch(`${process.env.OCL_CACHE}${urlIDCode}`, {
          signal: controller.signal,
        });
      } catch (e) {
        debug.fatal(e);
      }
      if (dataCompound?.status === 200) {
        success = true;
      } else {
        await delay(1000);
      }
      count++;
    }

    if (!success) {
      throw new Error('Failed to fetch data');
    }
    if (dataCompound?.status === 200) {
      let data = await dataCompound.json();
      let result = {
        data: {
          ocl: {
            idCode: data.result.idCode,
            coordinates: oclID.coordinates,
            noStereoTautomerID: data.result.noStereoTautomerID,

            acceptorCount: data.result.acceptorCount,
            donorCount: data.result.donorCount,
            logP: data.result.logP,
            logS: data.result.logS,
            polarSurfaceArea: data.result.polarSurfaceArea,
            rotatableBondCount: data.result.rotatableBondCount,
            stereoCenterCount: data.result.stereoCenterCount,
          },
          mf: data.result.mf,
          em: data.result.em,
          charge: data.result.charge,
          mw: data.result.mw,
          nbFragments: data.result.nbFragments,
          atoms: data.result.atoms,
          unsaturation: data.result.unsaturation,
        },
      };
      if (options.indexRequired) {
        result.data.ocl.index = data.result.ssIndex;
      }
      return result;
    } else {
      debug.fatal(`Error: ${dataCompound?.status} ${dataCompound}`);
    }
  } catch (e) {
    debug.fatal(e);
  }
}
