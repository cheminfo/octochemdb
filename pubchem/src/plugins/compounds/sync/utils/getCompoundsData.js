import MFParser from 'mf-parser';
import OCL from 'openchemlib';
import { getMF } from 'openchemlib-utils';

import Debug from '../../../../utils/Debug.js';

const { MF } = MFParser;
const debug = Debug('getCompoundsData');
/**
 * @description Calculate compounds properties (e.g. charge, OCL ID, molecular formula, etc.)
 * @param {*} molecule molecule from pubchem file
 * @returns {Object} compounds properties
 */
export function getCompoundsData(molecule) {
  let oclMolecule;
  if (molecule.molfile) {
    oclMolecule = OCL.Molecule.fromMolfile(molecule.molfile);
  } else {
    oclMolecule = OCL.Molecule.fromIDCode(molecule.noStereoID);
  }
  // calculate molecule properties (e.g. charge, OCL ID, molecular formula, etc.)
  const oclProperties = new OCL.MoleculeProperties(oclMolecule);
  const oclID = oclMolecule.getIDCodeAndCoordinates();
  const oclIndex = Array.from(oclMolecule.getIndex());
  const moleculeMF = getMF(oclMolecule);
  const mfParts = moleculeMF.parts;
  const nbFragments = mfParts.length;
  const mf = mfParts.join(' . ');
  const globalMF = moleculeMF.mf;
  oclMolecule.stripStereoInformation();
  const noStereoID = oclMolecule.getIDCode();
  let result = {
    data: {
      ocl: {
        idCode: oclID.idCode,
        coordinates: oclID.coordinates,
        index: oclIndex,
        noStereoID,

        acceptorCount: oclProperties.acceptorCount,
        donorCount: oclProperties.donorCount,
        logP: oclProperties.logP,
        logS: oclProperties.logS,
        polarSurfaceArea: oclProperties.polarSurfaceArea,
        rotatableBondCount: oclProperties.rotatableBondCount,
        stereoCenterCount: oclProperties.stereoCenterCount,
      },
      mf,
      nbFragments,
    },
  };

  try {
    // calculate molecular formula properties (ex. exact mass, unsaturations, etc.)
    const mfInfo = new MF(globalMF).getInfo();
    result.data.em = mfInfo.monoisotopicMass;
    result.data.mw = mfInfo.mass;
    result.data.unsaturation = mfInfo.unsaturation;
    result.data.charge = mfInfo.charge;
    result.data.atom = mfInfo.atoms;
  } catch (e) {
    debug(e);
  }
  return result;
}
