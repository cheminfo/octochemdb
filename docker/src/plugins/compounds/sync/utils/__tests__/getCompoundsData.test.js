import dotenv from 'dotenv';
import { test, expect } from 'vitest';

import { getCompoundsData } from '../getCompoundsData';

dotenv.config();

test('getCompoundsData', async () => {
  let molecule = {
    idCode:
      'ekTpA@@@LAEMGLn\\dTTRbRfLbteRrRTfbqbtRthdRjZFFfNnAQjjjjjjjfjjjjjijjh@@',
  };
  let dataCompound = await getCompoundsData(molecule);
  expect(dataCompound).toMatchSnapshot();
}, 10000);
