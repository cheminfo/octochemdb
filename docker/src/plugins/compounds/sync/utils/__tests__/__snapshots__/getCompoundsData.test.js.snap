// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getCompoundsData 1`] = `
{
  "data": {
    "atom": "{\\"C\\":37,\\"H\\":60,\\"O\\":8}",
    "charge": 0,
    "em": 632.42881889036,
    "mf": "C37H60O8",
    "mw": 632.868912817097,
    "nbFragments": 1,
    "ocl": {
      "acceptorCount": 8,
      "coordinates": "!BkDKv^?DDiHVdaYL}kDJlPoY{}gn[noDDiHVdaZqB^?D}kFslwoaF_zREQJYxN\\\\KL?Fr\\\\DfQAaLg|{OIKcTodY^s|aMXt|cSu?o\`t|o{xMOTt|cSrMOW~~@",
      "donorCount": 4,
      "idCode": "ekTpA@@@LAEMGLn\\\\dTTRbRfLbteRrRTfbqbtRthdRjZFFfNnAQjjjjjjjfjjjjjijjh@@",
      "index": [
        229,
        79,
        253,
        255,
        254,
        94,
        170,
        237,
        80,
        175,
        254,
        179,
        118,
        206,
        189,
        138,
        45,
        251,
        196,
        249,
        52,
        196,
        107,
        122,
        203,
        93,
        128,
        71,
        198,
        146,
        82,
        212,
        156,
        2,
        100,
        120,
        1,
        151,
        2,
        165,
        4,
        88,
        0,
        18,
        131,
        5,
        3,
        128,
        1,
        16,
        8,
        8,
        64,
        81,
        112,
        48,
        64,
        0,
        0,
        32,
        4,
        128,
        0,
        8,
      ],
      "logP": 4.544999875128269,
      "logS": -5.477999992668629,
      "noStereoTautomerID": "ekTpA@@@LAEMGLn\\\\dTTRbRfLbteRrRTfbqbtRthdRjZFFfNnAQjjjjjjjfjjjjjijjh@@",
      "polarSurfaceArea": 117.83999633789062,
      "rotatableBondCount": 8,
      "stereoCenterCount": 15,
    },
    "unsaturation": 8,
  },
}
`;
